proteuslib/unit_models/reverse_osmosis_1D.py

```diff
@@ -0,0 +1,800 @@
+        def eq_permeate_outlet_isothermal(b, t):
+            return b.feed_side.properties[t, 0].temperature == \
+                   b.permeate_out[t].temperature
+        # ==========================================================================
+        # isobaric conditions across permeate channel and at permeate outlet
+
+        @self.Constraint(self.flowsheet().config.time,
+                         self.feed_side.length_domain,
+                         doc="Isobaric assumption for permeate out")
+        def eq_permeate_outlet_isobaric(b, t, x):
+            return b.permeate_side[t, x].pressure == \
+                   b.permeate_out[t].pressure
+
+    def initialize(blk,
+                   feed_side_args=None,
+                   permeate_side_args=None,
+                   permeate_block_args=None,
+                   outlvl=idaeslog.NOTSET,
+                   solver=None,
+                   optarg=None):
+        """
+        Initialization routine for 1D-RO unit.
+
+        Keyword Arguments:
+            feed_side_args : a dict of arguments to be passed to the property
+             package(s) of the feed_side to provide an initial state for
+             initialization (see documentation of the specific
+             property package)
+            permeate_side_args : a dict of arguments to be passed to the property
+             package(s) of the permeate_side to provide an initial state for
+             initialization (see documentation of the specific
+             property package)
+            permeate_block_args : a dict of arguments to be passed to the property
+             package(s) of the final permeate StateBlock to provide an initial state for
+             initialization (see documentation of the specific
+             property package)
+            outlvl : sets output level of initialization routine
+            solver : str indicating which solver to use during
+                     initialization (default = None, use default solver)
+            optarg : solver options dictionary object (default=None, use default solver options)
+
+        Returns:
+            None
+        """
+        init_log = idaeslog.getInitLogger(blk.name, outlvl, tag="unit")
+        solve_log = idaeslog.getSolveLogger(blk.name, outlvl, tag="unit")
+
+        # Create solver
+        opt = get_solver(solver, optarg)
+
+        init_log.info('Starting initialization')
+        # ---------------------------------------------------------------------
+        # Step 1: Initialize feed_side, permeate_side, and permeate_out blocks
+        flags_feed_side = blk.feed_side.initialize(
+            outlvl=outlvl,
+            optarg=optarg,
+            solver=solver,
+            state_args=feed_side_args)
+
+        flags_permeate_side = blk.permeate_side.initialize(
+            outlvl=outlvl,
+            optarg=optarg,
+            solver=solver,
+            state_args=permeate_side_args)
```

Comment:
Extension for the future, but is there a way you could estimate some of these based on the feed state? Notably, once you have the permeate side, it is probably possible to directly calculate the permeate outlet state.

Reply:
Yes, I definitely would like to improve this. This is my first time writing an initialization routine like this, so still figuring it out things as I go. Would you happen to have any recommendations for resources on developing (getting started with developing) initialization routines? So far, I've just looked at several unit models in IDAES and followed along with the logic.

Reply:
Unfortunately, there are no good resources I am aware of. As far as examples go, the gas-solids contactors have some of the most complex initialisation routines, but they might also not be easy to follow.